### ISSUE TITLE
setBreakpoints is called with sourceModified:false when source is modified

### ISSUE THREAD

**Issue by DanTup** [microsoft/vscode#175872]
It's not clear from the DAP spec exactly what the intended behaviour is here, so I've filed https://github.com/microsoft/debug-adapter-protocol/issues/376. However, this part definitely seems like a bug.

I added a breakpoint to a file, modified the file, and then added another. At no point did I save the file.

![Mar-02-2023 11-08-38](https://user-images.githubusercontent.com/1078012/222412148-1f006a92-2abd-4a0b-a277-553102a0ccb0.gif)

In the DAP log, I see that VS Code called `setBreakpoints` after I added the second breakpoint but with `sourceModified: false`. This seems to be incorrect.. The debug adapter now thinks these breakpoints are for a saved file, but they are not.

```
[11:08:17] [DAP] [Info] ==> {"command":"setBreakpoints","arguments":{"source":{"name":"package:flutter_counter/main.dart","path":"/Users/danny/Desktop/dart_samples/flutter_counter/lib/main.dart","sourceReference":0},"lines":[14],"breakpoints":[{"line":14}],"sourceModified":false},"type":"request","seq":8}
[11:08:21] [DAP] [Info] ==> {"command":"setBreakpoints","arguments":{"source":{"name":"package:flutter_counter/main.dart","path":"/Users/danny/Desktop/dart_samples/flutter_counter/lib/main.dart","sourceReference":0},"lines":[14,15],"breakpoints":[{"line":14},{"line":15}],"sourceModified":false},"type":"request","seq":9}
```

**Comment by roblourens** [comment]
DAP says it "indicates that the underlying source has been modified __which results in new breakpoint locations.__"

So if you type above the breakpoints, pushing them down, I expect it to set sourceModified to true. Otherwise we are assuming that the original line numbers are still valid. Does that make sense?

**Comment by roblourens** [comment]
I think back then it worked the way you describe. Now it works the way I describe which I think matches DAP, I am not sure whether DAP changed.

**Comment by DanTup** [comment]
> which I think matches DAP

It's not clear to me what matching DAP means, because DAP only has that very vague line you quoted above. However it doesn't seem to match _either_ in my testing above, because it never sends `sourceModified: true` whether the file is dirty/unsaved, or was just modified since starting.

I think the spec needs clarifying since there are at least two possible interpretations, but VS code's current behaviour doesn't seem to match either (specifically, I cannot produce a case where VS Code sends `sourceModified: true`).

**Comment by roblourens** [comment]
I think we could make the DAP description a little more clear, but what I see is that if I type _above_ a breakpoint so that its original line number becomes invalid, then we send setBreakpoints with `sourceModified: true`

**Comment by DanTup** [comment]
hmm, that's not the behaviour I'm seeing (see the GIF and log above.. the log was captured while recording that GIF). Are there any conditions that should affect this?

If it is intended to work the way you describe (and we can figure out why it's not working for me), does Code additionally send a `setBreakpoints` request with `sourceModified: false` when the file is saved (and for subsequent breakpoint changes)? I really think this needs to be explicit in the spec - it's very difficult for DAs to do the right thing when things aren't documented explicitly (plus, Code behaviour often changes and it's not clear whether it's a Code bug or the DA is wrong if it's not explicit).

**Comment by roblourens** [comment]
Agreed about the spec being more explicit, but what I see in your gif is that your breakpoints are _above_ the place in the file that you changed, so their line numbers are considered to be the same, and I think that this is what `which results in new breakpoint locations` in the spec is describing.

**Comment by DanTup** [comment]
@roblourens this doesn't seem to make a difference. Here I added a breakpoint, modified the file _above it_ and then added another breakpoint, but I still get breakpoint requests with `sourceModified: true` and line numbers that do not match the file that the VM is running:

![Apr-03-2023 11-28-41](https://user-images.githubusercontent.com/1078012/229484721-7f22d4b6-d964-45a5-b6d3-b17c3e0faacb.gif)

```
[11:27:51] [DAP] [Info] ==> {"command":"setBreakpoints","arguments":{"source":{"name":"package:flutter_counter/main.dart","path":"/Users/danny/Desktop/dart_samples/flutter_counter/lib/main.dart","sourceReference":0},"lines":[18],"breakpoints":[{"line":18}],"sourceModified":false},"type":"request","seq":8}
[11:27:51] [DAP] [Info] <== {"seq":14789,"type":"response","request_seq":8,"command":"setBreakpoints","success":true,"body":{"breakpoints":[{"verified":true}]}}
[11:27:55] [DAP] [Info] ==> {"command":"setBreakpoints","arguments":{"source":{"name":"package:flutter_counter/main.dart","path":"/Users/danny/Desktop/dart_samples/flutter_counter/lib/main.dart","sourceReference":0},"lines":[14,20],"breakpoints":[{"line":14},{"line":20}],"sourceModified":false},"type":"request","seq":9}
[11:27:55] [DAP] [Info] <== {"seq":14799,"type":"response","request_seq":9,"command":"setBreakpoints","success":true,"body":{"breakpoints":[{"verified":true},{"verified":false}]}}
```

In the second request includes:

```
[{"line":14},{"line":20}],"sourceModified":false
```

But line 20 is not the line where this breakpoint was according to the VM, but VS Code has not given any indication to the DA that this file has been modified, so there's no way the DA can handle this. As far as it can tell, the user deleted the breakpoint on line 18 and added another on line 20.

I agree the spec definitely needs improving, but I also think VS Code's current behaviour doesn't make any sense either.

**Comment by roblourens** [comment]
Ok, yeah there is an issue with VS Code's behavior too. What I said before was partially correct, you should be getting an event with `sourceModified=true` once the file is saved. But that isn't remembered, and anything that triggers a normal setBreakpoints, like adding a new breakpoint, will send the event with `sourceModified=false` again. I need to look into this a bit.

In your scenario did you save the file?
